// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: core/user/user.proto

package user

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on UserInfo with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *UserInfo) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on UserInfo with the rules defined in
// the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in UserInfoMultiError, or nil
// if none found.
func (m *UserInfo) ValidateAll() error {
	return m.validate(true)
}

func (m *UserInfo) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	// no validation rules for Name

	// no validation rules for Email

	// no validation rules for ProfileImageUrl

	if len(errors) > 0 {
		return UserInfoMultiError(errors)
	}

	return nil
}

// UserInfoMultiError is an error wrapping multiple validation errors returned
// by UserInfo.ValidateAll() if the designated constraints aren't met.
type UserInfoMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m UserInfoMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m UserInfoMultiError) AllErrors() []error { return m }

// UserInfoValidationError is the validation error returned by
// UserInfo.Validate if the designated constraints aren't met.
type UserInfoValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e UserInfoValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e UserInfoValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e UserInfoValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e UserInfoValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e UserInfoValidationError) ErrorName() string { return "UserInfoValidationError" }

// Error satisfies the builtin error interface
func (e UserInfoValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sUserInfo.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = UserInfoValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = UserInfoValidationError{}

// Validate checks the field values on GetUserInfoByIdRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *GetUserInfoByIdRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on GetUserInfoByIdRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// GetUserInfoByIdRequestMultiError, or nil if none found.
func (m *GetUserInfoByIdRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *GetUserInfoByIdRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	if len(errors) > 0 {
		return GetUserInfoByIdRequestMultiError(errors)
	}

	return nil
}

// GetUserInfoByIdRequestMultiError is an error wrapping multiple validation
// errors returned by GetUserInfoByIdRequest.ValidateAll() if the designated
// constraints aren't met.
type GetUserInfoByIdRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m GetUserInfoByIdRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m GetUserInfoByIdRequestMultiError) AllErrors() []error { return m }

// GetUserInfoByIdRequestValidationError is the validation error returned by
// GetUserInfoByIdRequest.Validate if the designated constraints aren't met.
type GetUserInfoByIdRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GetUserInfoByIdRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GetUserInfoByIdRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GetUserInfoByIdRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GetUserInfoByIdRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GetUserInfoByIdRequestValidationError) ErrorName() string {
	return "GetUserInfoByIdRequestValidationError"
}

// Error satisfies the builtin error interface
func (e GetUserInfoByIdRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGetUserInfoByIdRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GetUserInfoByIdRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GetUserInfoByIdRequestValidationError{}

// Validate checks the field values on GetUserInfoByIdResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *GetUserInfoByIdResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on GetUserInfoByIdResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// GetUserInfoByIdResponseMultiError, or nil if none found.
func (m *GetUserInfoByIdResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *GetUserInfoByIdResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if m.UserInfo != nil {

		if all {
			switch v := interface{}(m.GetUserInfo()).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, GetUserInfoByIdResponseValidationError{
						field:  "UserInfo",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, GetUserInfoByIdResponseValidationError{
						field:  "UserInfo",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(m.GetUserInfo()).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return GetUserInfoByIdResponseValidationError{
					field:  "UserInfo",
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return GetUserInfoByIdResponseMultiError(errors)
	}

	return nil
}

// GetUserInfoByIdResponseMultiError is an error wrapping multiple validation
// errors returned by GetUserInfoByIdResponse.ValidateAll() if the designated
// constraints aren't met.
type GetUserInfoByIdResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m GetUserInfoByIdResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m GetUserInfoByIdResponseMultiError) AllErrors() []error { return m }

// GetUserInfoByIdResponseValidationError is the validation error returned by
// GetUserInfoByIdResponse.Validate if the designated constraints aren't met.
type GetUserInfoByIdResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GetUserInfoByIdResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GetUserInfoByIdResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GetUserInfoByIdResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GetUserInfoByIdResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GetUserInfoByIdResponseValidationError) ErrorName() string {
	return "GetUserInfoByIdResponseValidationError"
}

// Error satisfies the builtin error interface
func (e GetUserInfoByIdResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGetUserInfoByIdResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GetUserInfoByIdResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GetUserInfoByIdResponseValidationError{}
